packages/toolpad-app/src/toolpadDataSources/googleSheets/server.ts

```diff
@@ -43,45 +54,64 @@ function createOAuthClient() {
 
 async function execPrivate(
   connection: LegacyConnection<GoogleSheetsConnectionParams>,
-  query: GoogleSheetsQuery,
+  query: GoogleSheetsPrivateQuery,
 ): Promise<any> {
   const client = createOAuthClient();
   if (connection.params) {
     client.setCredentials(connection.params);
   }
-  if (query.type === GoogleSheetsActionType.FETCH_SHEET) {
-    const sheetsClient = google.sheets({
-      version: 'v4',
+  if (!query) {
+    return {};
+  }
+  if (query.type === GoogleSheetsPrivateQueryType.FILE_GET) {
+    const driveClient = google.drive({
+      version: 'v3',
       auth: client,
     });
-    const spreadsheetId = query.spreadsheet?.id;
-    try {
-      const response = await sheetsClient.spreadsheets.get({
-        spreadsheetId,
-        includeGridData: false,
-      });
-      if (response.statusText === 'OK') {
-        const { sheets } = response.data;
-        return { sheets: sheets?.map((sheet) => sheet.properties) };
-      }
-    } catch (error) {
-      throw new Error(`Unable to fetch spreadsheetId ${query.spreadsheet?.id}`);
+    const { spreadsheetId } = query;
+    const response = await driveClient.files.get({
+      fileId: spreadsheetId,
+    });
+    if (response.statusText === 'OK') {
+      return response.data;
     }
-  } else if (query.type === GoogleSheetsActionType.FETCH_SPREADSHEETS) {
+    throw new Error(`Failed to fetch query: "${JSON.stringify(query)}"`);
+  }
+  if (query.type === GoogleSheetsPrivateQueryType.FILES_LIST) {
     const driveClient = google.drive({
       version: 'v3',
       auth: client,
     });
-    try {
-      const response = await driveClient.files.list({
-        q: "mimeType='application/vnd.google-apps.spreadsheet'",
-      });
+    const { spreadsheetQuery, pageToken } = query;
+    let queryString = "mimeType='application/vnd.google-apps.spreadsheet'";
+    if (spreadsheetQuery) {
+      queryString = `name contains '${spreadsheetQuery}' and ${queryString}`;
+    }
+    const response = await driveClient.files.list({
+      q: queryString,
+      pageToken,
+    });
+    if (response.statusText === 'OK') {
+      return response.data;
+    }
+    throw new Error(`Failed to fetch query: "${JSON.stringify(query)}"`);
+  }
+  if (query.type === GoogleSheetsPrivateQueryType.FETCH_SPREADSHEET) {
+    const sheetsClient = google.sheets({
+      version: 'v4',
+      auth: client,
+    });
+    const { spreadsheetId } = query;
+    const response = await sheetsClient.spreadsheets.get({
+      spreadsheetId: spreadsheetId ?? undefined,
+      includeGridData: false,
+    });
+    if (response.statusText === 'OK') {
       return response.data;
-    } catch (error) {
-      throw new Error(`Unable to fetch spreadsheets`);
     }
+    throw new Error(`Failed to fetch query: "${JSON.stringify(query)}"`);
```

Comment:
Is there any information in the response that could help explain the error, that we can add to the error message? Like an error code or something?

Reply:
`response.status` contains a `number` with the HTTP Error code. We could do:

```js
throw new Error(`Error ${response.status}: Failed to fetch "${JSON.stringify(query)}"`
```